tests/test_validators.py

```diff
@@ -38,3 +39,44 @@ def test_number(self):
 
         with self.assertRaises(ValidationError):
             validator('spam')
+
+
+class ShapeTests(TestCase):
+
+    def test_shape_valid(self):
+        # Comma separated
+        self.assertEqual(is_shape('rect(1px, 3px, 2px, 4px)'), Rect(1, 3, 2, 4))
+
+        # Comma separated with extra spaces
+        self.assertEqual(is_shape(' rect( 1px,  3px,  2px,  4px ) '), Rect(1, 3, 2, 4))
+
+        # Space separated
+        self.assertEqual(is_shape('rect(1px 3px 2px 4px)'), Rect(1, 3, 2, 4))
+
+        # Space separated with extra spaces
+        self.assertEqual(is_shape(' rect( 1px  3px  2px  4px ) '), Rect(1, 3, 2, 4))
+
+    def test_shape_invalid(self):
+        # Mix of commas and spaces
+        with self.assertRaises(ValidationError):
+            is_shape('rect(1px, 3px, 2px 4px)')
+
+        # Incorrect number of elements
+        with self.assertRaises(ValidationError):
+            is_shape('rect(1px, 3px, 2px)')
+
+        with self.assertRaises(ValidationError):
+            is_shape('rect(1px, 3px)')
+
+        with self.assertRaises(ValidationError):
+            is_shape('rect(1px)')
+
+        with self.assertRaises(ValidationError):
+            is_shape('rect()')
+
+        # Other values
+        with self.assertRaises(ValidationError):
+            is_shape('(1px, 3px, 2px, 4px)')
+
+        with self.assertRaises(ValidationError):
+            is_shape('rect(a, b, c, d)')
```

Comment:
Some examples of "badly composed attempts at writing `rect()`" would be helpful here - e.g., mixing spaces and commas, missing open/close brace,...

Reply:
Move to use a regex 🤷‍♂